flocker/node/_deploy.py

```diff
@@ -934,6 +935,13 @@ def _restart_for_application_change(
         # we don't want to do anything:
         comparable_state = comparable_state.transform(["running"], True)
 
+        # Restart policies were briefly supported but they interact poorly with
```

Comment:
Not sure what I was thinking.  Possibly I got confused between which of the three variations of the solution I was working on. :/  I think your analysis is correct with one minor correction: I also changed `StartApplication` to start applications with `RestartNever()` as the policy regardless of configuration... So this change accidentally succeeds in only restarting the applications once.

The suggested fix makes the code make more sense.